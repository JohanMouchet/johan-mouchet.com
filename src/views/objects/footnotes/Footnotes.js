// @flow

import * as React from "react";
import cx from "classnames";
import "./footnotes.scss";

type ReferenceProps = {
  id: string,
  prefix: ?string,
  text: string,
  url: ?string
};

const FootnotesReference = (props: ReferenceProps) => {
  const { id, prefix, text, url } = props;

  return (
    <sup>
      <a
        id={`footnote-ref:${prefix && prefix + "-"}${id}`}
        className="o-footnote-ref"
        href={`${url}#footnote:${prefix && prefix + "-"}${id}`}
      >
        {text}
      </a>
    </sup>
  );
};

FootnotesReference.defaultProps = {
  url: ""
};

export { FootnotesReference };

type NotesProps = {
  notes: [
    {
      id: string,
      prefix: ?string,
      text: string,
      url: ?string
    }
  ],
  metadata: {
    contextClass: ?string,
    additionalClasses: ?Array<string>
  }
};

const FootnotesNotes = (props: NotesProps) => {
  const { notes, metadata } = props;

  const NotesClass = cx(
    "o-footnotes",
    metadata.contextClass,
    metadata.additionalClasses
  );

  return (
    <ol className={NotesClass}>
      {notes.map(note => (
        <li
          id={`footnote:${note.prefix && note.prefix + "-"}${note.id}`}
          className="o-footnotes__note"
        >
          {note.text}
          <a
            href={`${note.url}#footnote-ref:${note.prefix &&
              note.prefix + "-"}${note.id}`}
            className="footnote-backref"
          >
            &#x21a9;
          </a>
        </li>
      ))}
    </ol>
  );
};

FootnotesNotes.defaultProps = {
  url: ""
};

export { FootnotesNotes };
