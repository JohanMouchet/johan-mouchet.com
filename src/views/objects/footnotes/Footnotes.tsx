import React from "react";
import cx from "classnames";
import { IconArrowReturn } from "views/objects";
import styles from "./Footnotes.module.scss";

type ReferenceProps = {
  id: string;
  text: string;
  label?: string;
};

const FootnotesReference: React.FC<ReferenceProps> = ({
  id,
  text,
  label = "See note",
}) => (
  <sup>
    <a
      id={`footnote-ref:${id}`}
      className={styles["o-footnote-ref"]}
      href={`#footnote:${id}`}
      aria-label={label}
    >
      {text}
    </a>
  </sup>
);

export { FootnotesReference };

type NotesProps = {
  notes: Array<{
    id: string;
    text: string;
    label?: string;
  }>;
  className?: string | string[] | { [key: string]: boolean };
};

const FootnotesNotes: React.FC<NotesProps> = ({ notes, className }) => {
  const classNames = cx(styles["o-footnotes"], className);

  return !notes?.length ? null : (
    <ol className={classNames}>
      {notes.map((note) => (
        <li
          id={`footnote:${note.id}`}
          className={styles["o-footnotes__note"]}
          key={note.id}
        >
          {note.text}{" "}
          <a
            href={`#footnote-ref:${note.id}`}
            className={styles["o-footnotes__backref"]}
            aria-label={note.label || "Jump back to reference"}
          >
            <IconArrowReturn />
          </a>
        </li>
      ))}
    </ol>
  );
};

export { FootnotesNotes };
