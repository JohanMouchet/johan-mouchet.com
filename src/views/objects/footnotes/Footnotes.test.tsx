import { render } from "@testing-library/react";
import { expect, it } from "vitest";
import { FootnotesNotes, FootnotesReference } from "./Footnotes";

it("Renders <FootnotesReference /> without crashing", () => {
  expect(
    render(<FootnotesReference id="id-1" text="1" label="" />).container
      .firstChild
  ).toMatchSnapshot();
});

it("Renders <FootnotesNotes /> without crashing", () => {
  expect(
    render(
      <FootnotesNotes
        notes={[
          {
            id: "id-1",
            text: "Note",
            label: "",
          },
        ]}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});
