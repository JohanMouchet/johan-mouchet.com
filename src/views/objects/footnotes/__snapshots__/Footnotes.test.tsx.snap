// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders <FootnotesNotes /> without crashing 1`] = `
<ol
  class="_o-footnotes_dfc101"
>
  <li
    class="_o-footnotes__note_dfc101"
    id="footnote:id-1"
  >
    Note
     
    <a
      aria-label="Jump back to reference"
      class="_o-footnotes__backref_dfc101"
      href="#footnote-ref:id-1"
    >
      <svg
        viewBox="0 0 256 256"
      >
        <symbol
          fill="currentColor"
          id="6fdf6952"
          viewBox="0 0 256 256"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M232,56A104.11,104.11,0,0,1,128,160H51.31l34.35,34.34a8,8,0,0,1-11.32,11.32l-48-48a8,8,0,0,1,0-11.32l48-48a8,8,0,0,1,11.32,11.32L51.31,144H128a88.1,88.1,0,0,0,88-88,8,8,0,0,1,16,0Z"
          />
        </symbol>
        <use
          href="#6fdf6952"
        />
      </svg>
    </a>
  </li>
</ol>
`;

exports[`Renders <FootnotesReference /> without crashing 1`] = `
<sup>
  <a
    aria-label=""
    class="_o-footnote-ref_dfc101"
    href="#footnote:id-1"
    id="footnote-ref:id-1"
  >
    1
  </a>
</sup>
`;
